src/boot/efi/boot.c

```diff
@@ -1624,12 +1633,36 @@ static INTN config_entry_compare(const ConfigEntry *a, const ConfigEntry *b) {
         assert(a);
         assert(b);
 
-        /* Order entries that have no tries left to the beginning of the list */
-        if (a->tries_left != 0 && b->tries_left == 0)
-                return 1;
+        /* Order entries that have no tries left to the end of the list */
         if (a->tries_left == 0 && b->tries_left != 0)
+                return 1;
+        if (a->tries_left != 0 && b->tries_left == 0)
                 return -1;
 
+        /* First, order by sort key (alphabetically upwards), if it is defined. */
+        if (a->sort_key && b->sort_key) {
+                r = strcmp(a->sort_key, b->sort_key);
```

Comment:
`strcmp_ptr`?

Reply:
we already checked both args for non-NULL here, thus strcmp() is fine?

Reply:
You'd imho still want entries with sort-key to go before entries without one. Which this fails to do.

Reply:
I am not convinced that's desirable. I'd rather stick to the traditional sorting if we cannot compare by sort key.

Reply:
That create an "unstable" sorting order. Realistically, you cannot compare entries with to entries without. You're comparing apples vs oranges in that case, so ignoring the fact and resorting to comparing by color next makes no sense.

Reply:
To put this in more concrete terms, your sorting function is violating transitivity.

Given this list:
```
1: id=a.conf sort-key=50
2: id=b.conf
3: id=c.conf sort-key=40
```

You have these realtions:
```
1 < 2 (id)
2 < 3 (id)
```

Which by transitivity must imply `1 < 3`, but it doesn't because `1 > 3 (sort-key) !!!`

This means that depending on the original order of the list, you can generate an unstable sorting output.